src/dataset/impl_dataset.rs

```diff
@@ -724,50 +740,213 @@ where
+            let targets_sl = targets_sl2.as_slice_mut().unwrap();
+
+            for i in 0..k {
+                assist_swap_array2!(records_sl, i, fold_size, features);
+                assist_swap_array2!(targets_sl, i, fold_size, targets);
+
+                {
+                    let train = DatasetBase::new(
+                        ArrayView2::from_shape(
+                            (samples_count - fold_size, features),
+                            records_sl.split_at(fold_size * features).1,
+                        )
+                        .unwrap(),
+                        ArrayView2::from_shape(
+                            (samples_count - fold_size, targets),
+                            targets_sl.split_at(fold_size * targets).1,
+                        )
+                        .unwrap(),
+                    );
+
+                    let obj = fit_closure(&train);
+                    objs.push(obj);
+                }
+
+                assist_swap_array2!(records_sl, i, fold_size, features);
+                assist_swap_array2!(targets_sl, i, fold_size, targets);
+            }
         }
 
         objs.into_iter().zip(self.sample_chunks(fold_size))
-        //
     }
-}
 
-fn assist_swap_array2<F>(slice: &mut [F], index: usize, fold_size: usize, features: usize) {
-    if index == 0 {
-        return;
+    /// Cross validation for multi-target algorithms
+    ///
+    /// Given a list of fittable models, cross validation
+    /// is used to compare their performance according to some
+    /// performance metric. To do so, k-folding is applied to the
+    /// dataset and, for each fold, each model is trained on the training set
+    /// and its performance is evaluated on the validation set. The performances
+    /// collected for each model are then averaged over the number of folds.
+    ///
+    /// ### Parameters:
+    ///
+    /// - `k`: the number of folds to apply
+    /// - `parameters`: a list of models to compare
+    /// - `eval`: closure used to evaluate the performance of each trained model
+    ///
+    /// ### Returns
+    ///
+    /// An array of model performances, in the same order as the models in input, if no errors occur.
+    /// The performance of each model is given as an array of performances, one for each target.
+    /// Otherwise, it might return an Error in one of the following cases:
+    ///
+    /// - An error occurred during the fitting of one model
+    /// - An error occurred inside the evaluation closure
+    ///
+    /// ### Example
+    ///
+    /// ```rust, ignore
+    ///
+    /// use linfa::prelude::*;
+    ///
+    /// // mutability needed for fast cross validation
+    /// let mut dataset = linfa_datasets::diabetes();
+    ///
+    /// let models = vec![model1, model2, ... ];
+    ///
+    /// let r2_scores = dataset.cross_validate_mt(5,&models, |prediction, truth| prediction.r2(truth))?;
+    ///
+    /// ```
+    pub fn cross_validate_mt<O, DT, ER, M, FACC, C>(
```

Comment:
I still have to play around with the bounds but don't have enough time right now :/

Reply:
I pinged the ndarray maintainer on discord and maybe there's a way to get around the Array1/Array2 construction problem by collecting to vec and then using stack. I'll try and let you know how it goes

Reply:
I played around with the trait bounds and came up with this: https://github.com/Sauro98/linfa/pull/3/commits/ee5e368c33e5589b59fb91c72f25ca9cee9e4b3c

Reply:
it only works if the evaluation metric is the same for all targets though :thinking: perhaps we could rename the other to `cross_validate_multi` and use it for per-target metric function calls.

Reply:
Well, I like the idea of having a generic `cross_validate` which only allows you to get a summary over all the targets in case of multi target datasets, and then another one with a longer name that allows the user to keep targets separated.
This way one needs to use the less simple name only if they actually care about keeping the targets separated.

Also, by forcing the closure to take only array1s and then iterating over the prediction rows, you are bypassing the `MultiTargetRegression` trait ensuring no conflict when calling `prediction.metric(truth)` or `truth.metric(prediction)`